geopandas/array.py

```diff
@@ -988,34 +988,46 @@ def take(self, indices, allow_fill=False, fill_value=None):
         return GeometryArray(result, crs=self.crs)
 
     def _fill(self, idx, value):
-        """Fill index locations with value
+        """
+        Fill index locations with ``value``.
 
-        Value should be a BaseGeometry
+        ``value`` should be a BaseGeometry or a GeometryArray.
         """
-        if not (_is_scalar_geometry(value) or value is None):
+        if vectorized.isna(value):
+            value = [None]
+        elif _is_scalar_geometry(value):
+            value = [value]
+        elif isinstance(value, GeometryArray):
+            value = value[idx]
+        else:
             raise TypeError(
-                "Value should be either a BaseGeometry or None, got %s" % str(value)
+                "'value' parameter must be None, a scalar geometry, or a GeoSeries, "
+                f"but you passed a {type(value).__name__!r}"
             )
-        # self.data[idx] = value
-        value_arr = np.empty(1, dtype=object)
+
+        value_arr = np.empty(len(value), dtype=object)
         with compat.ignore_shapely2_warnings():
-            value_arr[:] = [value]
+            value_arr[:] = _shapely_to_geom(value)
+
         self.data[idx] = value_arr
         return self
 
     def fillna(self, value=None, method=None, limit=None):
-        """Fill NA/NaN values using the specified method.
+        """
+        Fill NA values with geometry (or geometries) or using the specified method.
 
         Parameters
         ----------
-        value : scalar, array-like
-            If a scalar value is passed it is used to fill all missing values.
-            Alternatively, an array-like 'value' can be given. It's expected
-            that the array-like have the same length as 'self'.
+        value : pygeos.Geometry, shapely geometry object or GeometryArray
```

Comment:
```suggestion
        value : scalar geometry object or GeometryArray
```
Again, less we mention pygeos the better.